diff --git a/dlls/ntdll/ntdll.spec b/dlls/ntdll/ntdll.spec
index 32b1d5e..fc4c4ca 100644
--- a/dlls/ntdll/ntdll.spec
+++ b/dlls/ntdll/ntdll.spec
@@ -1051,6 +1051,7 @@
 @ stub RtlStartRXact
 # @ stub RtlStatMemoryStream
 @ stdcall RtlStringFromGUID(ptr ptr)
+@ stdcall RtlStringFromGUIDEx(ptr ptr long)
 @ stdcall RtlSubAuthorityCountSid(ptr)
 @ stdcall RtlSubAuthoritySid(ptr long)
 @ stdcall RtlSubtreePredecessor(ptr)
diff --git a/dlls/ntdll/rtlstr.c b/dlls/ntdll/rtlstr.c
index f494d97..f3878a5 100644
--- a/dlls/ntdll/rtlstr.c
+++ b/dlls/ntdll/rtlstr.c
@@ -1703,6 +1703,47 @@ NTSTATUS WINAPI RtlStringFromGUID(const GUID* guid, UNICODE_STRING *str)
   return STATUS_SUCCESS;
 }
 
+/*************************************************************************
+ * RtlStringFromGUIDEx (NTDLL.@)
+ *
+ * Convert a GUID into a string representation of a GUID.
+ *
+ * PARAMS
+ *  guid     [I] GUID to convert
+ *  str      [O] Destination for the converted string
+ *  allocate [I] Allocate str->Buffer from the process heap
+ *
+ * RETURNS
+ *  Success: STATUS_SUCCESS. str contains the converted value.
+ *  Failure: STATUS_NO_MEMORY if allocation fails,
+ *           STATUS_BUFFER_TOO_SMALL if the supplied buffer is too small.
+ */
+NTSTATUS WINAPI RtlStringFromGUIDEx(const GUID* guid, UNICODE_STRING *str, BOOLEAN allocate)
+{
+  TRACE("(%p,%p,%u)\n", guid, str, allocate);
+
+  str->Length = GUID_STRING_LENGTH * sizeof(WCHAR);
+  if (allocate)
+  {
+    str->MaximumLength = str->Length + sizeof(WCHAR);
+    str->Buffer = RtlAllocateHeap(GetProcessHeap(), 0, str->MaximumLength);
+    if (!str->Buffer)
+    {
+      str->Length = str->MaximumLength = 0;
+      return STATUS_NO_MEMORY;
+    }
+  }
+  else if (str->MaximumLength < str->Length + sizeof(WCHAR))
+    return STATUS_BUFFER_TOO_SMALL;
+
+  swprintf(str->Buffer, str->MaximumLength/sizeof(WCHAR),
+          L"{%08lX-%04X-%04X-%02X%02X-%02X%02X%02X%02X%02X%02X}", guid->Data1, guid->Data2, guid->Data3,
+          guid->Data4[0], guid->Data4[1], guid->Data4[2], guid->Data4[3],
+          guid->Data4[4], guid->Data4[5], guid->Data4[6], guid->Data4[7]);
+
+  return STATUS_SUCCESS;
+}
+
 
 /***********************************************************************
  * Message formatting
